Add entrainment rule BLD calculation to KPP

gustavo-marques opened the pull request:
This commit introduces an alternative method for computing the boundary layer depth (BLD) in KPP, based on the entrainment rule. This method is activated when the non-solar surface buoyancy flux (`surfBuoy_NS`) is negative and `STOKES_MOST = True`. In all other cases, the standard Richardson Number-based method is used.

The non-solar surface buoyancy flux is estimated internally using an exponential attenuation function, since only the total surface buoyancy flux (solar + non-solar) is currently available to the module. Ideally, the non-solar component should be provided explicitly.

Several new diagnostics have been added, and some of the existing ones have been updated. Specifically, the parameterized shear, buoyancy, and Stokes TKE production terms are calculated. We conducted multiple tests where these terms are passed to the MLE module and used in the denominator of the Bodner streamfunction. However, because of the small Cr value we adopted (0.01), this had minimal effect on the solutions, so we decided to keep the original simpler implementation and not pursue this option for now.

This commit should be evaluated together with https://github.com/CVMix/CVMix-src/pull/106.

gustavo-marques committed:
Add entrainment rule BLD calculation to KPP

This commit introduces an alternative method for computing the boundary layer depth
(BLD) in KPP, based on the entrainment rule. This method is activated when the
non-solar surface buoyancy flux (surfBuoy_NS) is negative and STOKES_MOST = True.
In all other cases, the standard Richardson Number-based method is used.

The non-solar surface buoyancy flux is estimated internally using an exponential
attenuation function, since only the total surface buoyancy flux (solar + non-solar)
is currently available to the module. Ideally, the non-solar component should be
provided explicitly.

Several new diagnostics have been added, and some of the existing ones have been
updated. Specifically, the parameterized shear, buoyancy, and Stokes TKE production
terms are calculated. We conducted multiple tests where these terms are passed to
the MLE module and used in the denominator of the Bodner streamfunction.
However, because of the small Cr value we adopted (0.01), this had minimal effect
on the solutions, so we decided to keep the original simpler implementation
and not pursue this option for now.

This commit should be evaluated together with https://github.com/CVMix/CVMix-src/pull/106.

gustavo-marques committed:
Fix typos

gustavo-marques committed:
Update CVMix to latest commit

Include new entreinment rule depth changes.

gustavo-marques commented on src/parameterizations/vertical/MOM_CVMix_KPP.F90:
```diff
@@ -548,7 +560,7 @@ logical function KPP_init(paramFile, G, GV, US, diag, Time, CS, passive)
                        CVMix_kpp_params_user=CS%KPP_params )
 
-  ! Register diagnostics
+  ! Redgister diagnostics
```
Fixed.

gustavo-marques commented on src/parameterizations/vertical/MOM_CVMix_KPP.F90:
```diff
@@ -1066,18 +1098,22 @@ subroutine KPP_compute_BLD(CS, G, GV, US, h, Temp, Salt, u, v, tv, uStar, buoyFl
+  real, dimension( GV%ke )   :: StokesXI_1d ! Parameters of TKE production ratio [nondim]
+  real, dimension( GV%ke )   :: BEdE_ER_1d  ! Entrainment Rule parameterized  [Z^3 T-3 ~> m s-1]
+  real :: ERdepth ! Entrainment Rule Boundary layer depth  CVMix_kpp_compute_ER_depth in MKS units [m]
+  real :: check ! Entrainmnent Rule Boundary layer depth  CVMix_kpp_compute_ER_depth in MKS units [m]
```
Fixed.

gustavo-marques committed:
Revert unit conversion for Vt2 and GoRho

gustavo-marques commented on src/parameterizations/vertical/MOM_CVMix_KPP.F90:
```diff
@@ -1092,11 +1128,11 @@ subroutine KPP_compute_BLD(CS, G, GV, US, h, Temp, Salt, u, v, tv, uStar, buoyFl
 
   ! some constants
-  GoRho = US%Z_to_m*US%s_to_T**2 * (GV%g_Earth_Z_T2 / GV%Rho0)
+  GoRho = US%Z_to_m*US%s_to_T**2 * (US%L_to_Z**2 * GV%g_Earth / GV%Rho0)
```
Reverted GoRho to the previous version.

gustavo-marques commented on src/parameterizations/vertical/MOM_CVMix_KPP.F90:
```diff
@@ -1459,40 +1536,37 @@ subroutine KPP_compute_BLD(CS, G, GV, US, h, Temp, Salt, u, v, tv, uStar, buoyFl
                     uStar=surfFricVel,  & ! surface friction velocity [m s-1]
                     CVmix_kpp_params_user=CS%KPP_params ) ! KPP parameters
-        CS%Vt2(i,j,:) = US%m_to_Z**2*US%T_to_s**2 * Vt2_1d(:)
+        if (CS%id_Vt2 > 0) CS%Vt2(i,j,:) = US%m_to_Z*US%T_to_s * Vt2_1d(:)
```
Fixed.

gustavo-marques committed:
Update description and remove empty line

mnlevy1981 commented on src/parameterizations/vertical/MOM_CVMix_KPP.F90:
```diff
@@ -1763,47 +1840,69 @@ subroutine KPP_NonLocalTransport_saln(CS, G, GV, h, nonLocalTrans, surfFlux, dt,
+  real,                intent(inout) :: uS_k , vS_k     !< Stokes velocity at zbl               [m s-1]
+  real,                intent(inout) :: uS_SL, vS_SL    !< Stokes velocity at zSL               [m s-1]
+  real,                intent(inout) :: uSbar, vSbar    !< mean Stokes velocity (ztop to zbot)  [m s-1]
+  real,                intent(inout) :: uSb_SL, vSb_SL  !< mean Stokes velocity (zSLtop to zSL) [m s-1]
```
I think the `check-style-and-docstrings` test is failing because you have multiple variables per line, this needs to be split to 15 lines instead of 7

gustavo-marques replied:
Done. Thanks, Mike!

gustavo-marques committed:
Split description  of variables

gustavo-marques committed:
Merge branch 'dev/ncar' into kpp_ER_depth

gustavo-marques committed:
Refactor KPP StokesMOST diagnostics and allocations

- Removed unused `ustar` array from KPP_CS type.
- Restricted registration, allocation, and posting of StokesMOST
  diagnostics (`StokesXI`, `Lam2`, `BEdE_ER`, `ERdepth`, `RNdepth`,
  `PU_TKE`, `PS_TKE`, `PB_TKE`) to cases when `CS%StokesMOST` is enabled.
- Moved allocations of related arrays inside StokesMOST conditional block.
- Updated condition for calculating `surfBuoy_NS` to avoid division
  by zero in alog(buoyFlux(i,j,2)/buoyFlux(i,j,3)).
- Wrapped diagnostic and debug checks in `if (CS%StokesMOST)` conditions.
- Minor cleanup of comments and diagnostic labeling.

gustavo-marques commented:
This PR introduces expected answer changes for `pr_mom`; all other tests pass successfully.

/glade/derecho/scratch/gmarques/cesm.tests/pr_mom/alpha07c_dev_ncar--kpp_ER_depth.intel

```
  DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods (Overall: DIFF) details:
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods CREATE_NEWCASE
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods XML
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods SETUP
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods SHAREDLIB_BUILD time=43
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods NLCOMP
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods MODEL_BUILD time=145
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods SUBMIT
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods RUN time=4552
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods COMPARE_dimscale_1_Tp_base
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods COMPARE_dimscale_2_Lp_base
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods COMPARE_dimscale_3_Hp_base
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods COMPARE_dimscale_4_Zp_base
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods COMPARE_dimscale_5_Rp_base
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods COMPARE_dimscale_6_Qp_base
    FAIL DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods BASELINE alpha07c_dev_ncar--dev_ncar: DIFF
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods TPUTCOMP
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods MEMLEAK insufficient data for memleak test
    PASS DIMCSL_Ld1.TL319_t232.G_JRA.derecho_intel.mom-cfc_mods SHORT_TERM_ARCHIVER
  ERI.TL319_t232.G_JRA.derecho_intel.mom-debug (Overall: DIFF) details:
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug CREATE_NEWCASE
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug XML
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug SETUP
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug SHAREDLIB_BUILD time=10
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug NLCOMP
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug MODEL_BUILD time=96
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug SUBMIT
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug RUN time=2715
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug COMPARE_base_hybrid
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug COMPARE_base_rest
    FAIL ERI.TL319_t232.G_JRA.derecho_intel.mom-debug BASELINE alpha07c_dev_ncar--dev_ncar: DIFF
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug TPUTCOMP
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug MEMLEAK insufficient data for memleak test
    PASS ERI.TL319_t232.G_JRA.derecho_intel.mom-debug SHORT_TERM_ARCHIVER
  ERS.TL319_t232_wg37.GW_JRA.derecho_intel (Overall: DIFF) details:
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel CREATE_NEWCASE
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel XML
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel SETUP
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel SHAREDLIB_BUILD time=9
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel NLCOMP
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel MODEL_BUILD time=310
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel SUBMIT
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel RUN time=965
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel COMPARE_base_rest
    FAIL ERS.TL319_t232_wg37.GW_JRA.derecho_intel BASELINE alpha07c_dev_ncar--dev_ncar: DIFF
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel MEMCOMP
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel TPUTCOMP
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel MEMLEAK
    PASS ERS.TL319_t232_wg37.GW_JRA.derecho_intel SHORT_TERM_ARCHIVER
  SMS_D.TL319_t232.G_JRA_RYF.derecho_intel (Overall: DIFF) details:
    PASS SMS_D.TL319_t232.G_JRA_RYF.derecho_intel CREATE_NEWCASE
    PASS SMS_D.TL319_t232.G_JRA_RYF.derecho_intel XML
    PASS SMS_D.TL319_t232.G_JRA_RYF.derecho_intel SETUP
    PASS SMS_D.TL319_t232.G_JRA_RYF.derecho_intel SHAREDLIB_BUILD time=9
    PASS SMS_D.TL319_t232.G_JRA_RYF.derecho_intel NLCOMP
    PASS SMS_D.TL319_t232.G_JRA_RYF.derecho_intel MODEL_BUILD time=56
    PASS SMS_D.TL319_t232.G_JRA_RYF.derecho_intel SUBMIT
    PASS SMS_D.TL319_t232.G_JRA_RYF.derecho_intel RUN time=1054
    FAIL SMS_D.TL319_t232.G_JRA_RYF.derecho_intel BASELINE alpha07c_dev_ncar--dev_ncar: DIFF
    PASS SMS_D.TL319_t232.G_JRA_RYF.derecho_intel MEMCOMP
    PASS SMS_D.TL319_t232.G_JRA_RYF.derecho_intel TPUTCOMP
    PASS SMS_D.TL319_t232.G_JRA_RYF.derecho_intel MEMLEAK
    PASS SMS_D.TL319_t232.G_JRA_RYF.derecho_intel SHORT_TERM_ARCHIVER
  SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel (Overall: DIFF) details:
    PASS SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel CREATE_NEWCASE
    PASS SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel XML
    PASS SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel SETUP
    PASS SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel SHAREDLIB_BUILD time=139
    PASS SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel NLCOMP
    PASS SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel MODEL_BUILD time=321
    PASS SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel SUBMIT
    PASS SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel RUN time=402
    FAIL SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel BASELINE alpha07c_dev_ncar--dev_ncar: DIFF
    PASS SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel TPUTCOMP
    PASS SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel MEMLEAK insufficient data for memleak test
    PASS SMS_Ld2.ne30pg3_t232.B1850C_LTso.derecho_intel SHORT_TERM_ARCHIVER
```
